File: 3562.maximum-profit-from-trading-stocks-with-discounts.cpp
#
# @lc app=leetcode id=3562 lang=cpp
#
# [3562] Maximum Profit from Trading Stocks with Discounts
#
# @lc code=start
class Solution {
public:
    int maxProfit(int n, vector<int>& present, vector<int>& future, vector<vector<int>>& hierarchy, int budget) {
        // Build adjacency list for tree
        vector<vector<int>> children(n + 1);
        for (auto& edge : hierarchy) {
            children[edge[0]].push_back(edge[1]);
        }
        
        // Memoization: (node, budgetLeft, parentBought) -> maxProfit
        map<tuple<int, int, int>, int> memo;
        
        function<int(int, int, int)> dfs = [&](int node, int budgetLeft, int parentBought) -> int {
            auto key = make_tuple(node, budgetLeft, parentBought);
            if (memo.count(key)) return memo[key];
            
            // Calculate cost and profit for current node
            int cost = parentBought ? present[node - 1] / 2 : present[node - 1];
            int profit = future[node - 1] - cost;
            
            // Option 1: Don't buy current node's stock
            // Use knapsack DP to allocate budget to children (no discount)
            vector<int> dpNoBuy(budgetLeft + 1, 0);
            for (int child : children[node]) {
                vector<int> newDp(budgetLeft + 1, 0);
                for (int b = 0; b <= budgetLeft; b++) {
                    newDp[b] = dpNoBuy[b]; // Don't allocate to this child
                    for (int cb = 1; cb <= b; cb++) {
                        newDp[b] = max(newDp[b], dpNoBuy[b - cb] + dfs(child, cb, 0));
                    }
                }
                dpNoBuy = newDp;
            }
            
            int maxProfit = dpNoBuy[budgetLeft];
            
            // Option 2: Buy current node's stock
            if (cost <= budgetLeft) {
                int remaining = budgetLeft - cost;
                // Use knapsack DP to allocate remaining budget to children (with discount)
                vector<int> dpBuy(remaining + 1, 0);
                for (int child : children[node]) {
                    vector<int> newDp(remaining + 1, 0);
                    for (int b = 0; b <= remaining; b++) {
                        newDp[b] = dpBuy[b];
                        for (int cb = 1; cb <= b; cb++) {
                            newDp[b] = max(newDp[b], dpBuy[b - cb] + dfs(child, cb, 1));
                        }
                    }
                    dpBuy = newDp;
                }
                maxProfit = max(maxProfit, profit + dpBuy[remaining]);
            }
            
            return memo[key] = maxProfit;
        };
        
        return dfs(1, budget, 0);
    }
};
# @lc code=end

File: 3562.maximum-profit-from-trading-stocks-with-discounts.py
#
# @lc app=leetcode id=3562 lang=python3
#
# [3562] Maximum Profit from Trading Stocks with Discounts
#

# @lc code=start
from typing import List

class Solution:
    def maxProfit(self, n: int, present: List[int], future: List[int], hierarchy: List[List[int]], budget: int) -> int:
        present = [0] + present
        future = [0] + future
        adj = [[] for _ in range(n + 1)]
        for a, b in hierarchy:
            adj[a].append(b)
        INF = 10**9 + 7
        memo = {}

        def dfs(u: int, disc: int):
            key = (u, disc)
            if key in memo:
                return memo[key]
            cost = present[u] // 2 if disc else present[u]
            prof = future[u] - cost
            children = adj[u]
            comb0 = combine(children, 0)
            comb1 = combine(children, 1)
            dp_no = comb0[:]
            dp_yes = [-INF] * (budget + 1)
            for k in range(budget + 1):
                if k >= cost:
                    dp_yes[k] = prof + comb1[k - cost]
            memo[key] = (dp_no, dp_yes)
            return memo[key]

        def combine(children: list[int], dchild: int):
            B = budget
            comb = [0] * (B + 1)
            for v in children:
                c_no, c_yes = dfs(v, dchild)
                child_max = [0] * (B + 1)
                for j in range(B + 1):
                    mx = c_no[j]
                    if c_yes[j] > -INF:
                        mx = max(mx, c_yes[j])
                    child_max[j] = mx
                new_comb = [-INF] * (B + 1)
                for p in range(B + 1):
                    for m in range(B - p + 1):
                        new_comb[p + m] = max(new_comb[p + m], comb[p] + child_max[m])
                for j in range(1, B + 1):
                    new_comb[j] = max(new_comb[j], new_comb[j - 1])
                comb = new_comb
            return comb

        if n == 0:
            return 0
        dp_no, dp_yes = dfs(1, 0)
        ans = 0
        for k in range(budget + 1):
            ans = max(ans, dp_no[k])
            if dp_yes[k] > -INF:
                ans = max(ans, dp_yes[k])
        return ans
# @lc code=end